library/nist_kw.c

```diff
@@ -0,0 +1,650 @@
+                            const unsigned char *input, size_t in_len,
+                            unsigned char *output, size_t* out_len )
+{
+    int ret = 0;
+    size_t i;
+    const uint32_t ICV1[] = NIST_KW_ICV1;
+    const uint32_t ICV2[] = NIST_KW_ICV2;
+
+    if( mode == MBEDTLS_KW_MODE_KW )
+    {
+        /*
+         * According to SP800-38F Table 1, the ciphertext length for KW
+         * must be between 3 to 2^54 semiblocks inclusive.
+         */
+        if( ( in_len < 24 )
+#if SIZE_MAX > UINT_MAX
+            || ( in_len > 0x200000000000000 )
+#endif
+            )
+        {
+            return ( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+        }
+
+        ret = unwrap( ctx, input, in_len / KW_SEMIBLOCK_LENGTH, output, out_len );
+        if( ret != 0 )
+            return ( ret );
+
+        if ( memcmp( ICV1, output, KW_SEMIBLOCK_LENGTH ) != 0 )
+        {
+            ret = MBEDTLS_ERR_CIPHER_AUTH_FAILED;
+            goto cleanup;
+        }
+
+        /*
+         * shift output to point to P
+         */
+        for( i = 0; i < ( *out_len / KW_SEMIBLOCK_LENGTH ); i++ )
+        {
+            memcpy( output + (i * KW_SEMIBLOCK_LENGTH), output + ( ( i + 1 ) * KW_SEMIBLOCK_LENGTH), KW_SEMIBLOCK_LENGTH );
+        }
+        *out_len = *out_len - KW_SEMIBLOCK_LENGTH;
+    }
+    else // KWP
+    {
+        int padlen = 0;
+        uint32_t Plen;
+        /*
+         * According to SP800-38F Table 1, the ciphertext length for KWP
+         * must be between 2 to 2^29 semiblocks inclusive.
+         */
+        if( ( in_len < KW_SEMIBLOCK_LENGTH * 2 )
+#if SIZE_MAX > UINT_MAX
+         || ( in_len > 0x100000000 )
+#endif
+        )
+        {
+            return ( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+        }
+
+        if( in_len == KW_SEMIBLOCK_LENGTH * 2 )
+        {
+            ret = mbedtls_cipher_update( &ctx->cipher_ctx, input, 16, output, out_len );
+            if( ret != 0 )
+                goto cleanup;
+        }
+        else if( in_len >  KW_SEMIBLOCK_LENGTH * 2 )
+        {
+            ret = unwrap( ctx, input, in_len / KW_SEMIBLOCK_LENGTH, output, out_len );
+            if( ret != 0 )
+                return ( ret );
+        }
+        if( memcmp( ICV2, output, KW_SEMIBLOCK_LENGTH / 2 ) != 0 )
+        {
+            ret = MBEDTLS_ERR_CIPHER_AUTH_FAILED;
+            goto cleanup;
+        }
+
+        GET_UINT32_BE( Plen, output, KW_SEMIBLOCK_LENGTH / 2 );
+
+        padlen = 8 * ( ( in_len / 8 ) - 1 ) - Plen;
+        if ( padlen < 0 || padlen > 7 )
+        {
+            ret = MBEDTLS_ERR_CIPHER_AUTH_FAILED;
+            goto cleanup;
+        }
+        for( i = 0; i < (size_t)padlen; i++ )
+        {
+            if( output[ in_len - padlen + i ] != 0 )
+            {
+                ret = MBEDTLS_ERR_CIPHER_AUTH_FAILED;
+                goto cleanup;
+            }
+        }
+        /*
+         * shift output to point to P
+         */
+
```

Comment:
Remove this empty newline between the comment and the code that it applies to. Consider adding a newline above the comment instead.